src/duration.cpp

```diff
@@ -0,0 +1,207 @@
+#include <regex>
+#include <stdexcept>
+#include <tuple>
+#include <limits>
+#include <Rcpp.h>
+#include "duration.hpp"
+#include "pseudovector.hpp"
+#include "utilities.hpp"
+
+
+Global::duration from_string(const std::string& str) {
+  Global::duration d = std::chrono::seconds(0);
+  const char* s = str.c_str();
+  const char* e = s + str.size();
+
+  auto sign = 1;
+  if (s < e && *s == '-') {
+    sign = -1;
+    ++s;
+  }
+
+  int n;
+  if (!Global::readNumber(s, e, n, false)) throw std::range_error("cannot parse nanoduration");
+
+  if (s < e && *s == ':') {
+    // we've got HHH:MM:SS format
+    d += n * std::chrono::hours(1);
+    ++s;
+    if (s + 5 > e || !isdigit(s[0]) || !isdigit(s[1]) || 
+        s[2] != ':' || !isdigit(s[3]) || !isdigit(s[4])) {
+      throw std::range_error("cannot parse nanoduration");
+    }
+    d += ((s[0] - '0')*10 + (s[1] - '0'))*std::chrono::minutes(1);
+
+    // treat seconds in the general way:
+    n = (s[3] - '0')*10 + (s[4] - '0');
+    s += 5;
+  }
+
+  d += n * std::chrono::seconds(1);
+  if (s == e) return sign*d;
+
+  if (*s++ != '.') throw std::range_error("cannot parse nanoduration");
+  Global::duration mul = std::chrono::milliseconds(100);
+  unsigned i = 0;
+  while (s < e) {
+    if (mul < std::chrono::nanoseconds(1)) throw std::range_error("cannot parse nanoduration");
+    if ((i == 3 || i == 6) && *s == '_') { ++s; continue; }
+    ++i;
+    if (!isdigit(*s)) throw std::range_error("cannot parse nanoduration");
+    d += (*s - '0') * mul;
+    mul /= 10;
+    ++s;
+  }
+
+  return sign*d;
+}
+
+
+std::string to_string(Global::duration d) {
+  std::stringstream ss;
+
+  if (is_na(d)) {
+    ss << "";
+  }
+  else {
+    // handle hh:mm:ss
+    if (d < std::chrono::seconds(0)) {
+      ss << '-';
+      d *= -1;
+    }
+    auto h = d / std::chrono::hours(1);
+    d -= h * std::chrono::seconds(3600);
+    auto min = d / std::chrono::minutes(1);
+    d -= min * std::chrono::seconds(60);
+    auto s = d / std::chrono::seconds(1);
+    d -= s * std::chrono::seconds(1);
+    ss << std::setfill ('0') 
+       << std::setw(2) << h  << ':' 
+       << std::setw(2) << min  << ':' 
+       << std::setw(2) << s;
+  
+    // now handle nanoseconds 000_000_000
+    auto ms = d / std::chrono::milliseconds(1);
+    d -= ms * std::chrono::milliseconds(1);
+    auto us = d / std::chrono::microseconds(1);
+    d -= us * std::chrono::microseconds(1);
+    auto ns = d / std::chrono::nanoseconds(1);
+    d -= ns * std::chrono::nanoseconds(1);
+
+    if (ms || us || ns) {
+      ss << '.';
+      ss << std::setfill ('0') << std::setw(3) << ms;
+      if (us || ns) {
+        ss << '_' << std::setfill('0') << std::setw(3) << us;
+        if (ns) {
+          ss << '_' << std::setfill('0') << std::setw(3) << ns;
+        }
+      }
+    }
+  }
+  
+  return ss.str();
+}
+
+
+bool is_na(Global::duration d) {
+  return d == Global::duration::min();
+}
+
+
+RcppExport SEXP duration_from_string(SEXP s) {
+  try {
```

Comment:
We can do possibly one better here by still relying on `// [[Rcpp::export]]` to get automatic type conversion AND the `try/catch` wrapper.  One trick is to use `// [[Rcpp::export(.duration_from_string)]]` which does all the work for us, but still does not make the function 'public' in the "need a manual page for it too" sense.

Reply:
Ah, great, didn't know that. That is better and safer. Will take a look at that.